physics/forkable.hpp

```diff
@@ -118,6 +118,24 @@ class Forkable {
   // |trajectory|.
   void DeleteFork(Tr4jectory*& trajectory);
 
+  // Prepends |trajectory| to this trajectory.  The algorithm proceeds as
+  // follows, where t is the start of the timeline of this trajectory:
+  // 1. Locate the most-forked timeline of |trajectory| that contains a point
+  //    (strictly) before t.  Call the trajectory containing that timeline T.
+  // 2. Prepend the beginning (strictly before t) of the timeline of T to the
+  //    timeline of this trajectory.
+  // 3. Attach the forks (strictly) before t to this trajectory.
+  // 4. If T is a root, delete the entire tree rooted at T.
+  // 5. Otherwise, attach this trajectory as a fork of the parent of T instead
+  //    of T and delete T.
+  // When returning from this method, it is generally not safe to touch any part
+  // of the tree containing |trajectory|, unless the client knows a lot about
+  // the structure of the various trajectories (that's the case in tests).
+  // This trajectory is still a valid pointer, but it may now be owned by the
+  // parent of T.  The only pointer that the client might assume to own is
+  // this->root().
```

Comment:
Let’s make this a static method that takes both roots by `unique_ptr` as well as a pointer to `trajectory` (whose root must be the prepended one), and returns whichever `unique_ptr` remains; taking ownership of `this` is a bad idea (and likewise for `trajectory->root()`), and having a `unique_ptr` to the non-const root of `trajectory` will ensure that the `const_cast` in the implementation does not introduce UB.